### PR TITLE
Convert reference layer selection to bottom sheet in select one from map

### LINKED ISSUE #5845: Convert reference layer selection to bottom sheet in select one from map
## Acceptance

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
And no layer has been selected before
When I open the select one from map view
And I press the layers button
Then I see a list of the available layers as bottom sheet
And "None" is selected

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
And I've selected one of those layers in settings
When I open the select one from map view
And I press the layers button
Then I see a list of the available layers as bottom sheet
And the expected layer is selected

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
When I open the select one from map view
And I press the layers button
And I swipe the bottom sheet up above the middle of the screen
Then the bottom sheet should fill the screen

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
When I open the select one from map view
And I press the layers button
And I press back
Then I should return the map

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
When I open the select one from map view
And I press the layers button
And I select a layer
And I click "Save"
Then I am returned to the map with that layer being used

- [x] Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
When I open the select one from map view
And I press the layers button
And I select a layer
And I click "Cancel"
Then I am returned to the map with whatever was being used before as the reference layer

## Notes

- Figma link: https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows.
- In the current/old UI, the list of layers are loaded on the UI thread (straight from `DirectoryReferenceLayerRepository`). Given this involves scanning the file system, this should definitely be happening in a background thread (probably using `Scheduler#immediate`) and the bottom sheet should display a centred (in the bottom sheet) [circular progress indicator](https://m3.material.io/components/progress-indicators/overview) while they're loading.
- The implementation of the bottom sheet itself will probably be similar to the summary sheet in `SelectionMapFragment` although it might need to be a `Fragment` itself so that it's more reusable (as it will eventually be used in other places).

### LINKED ISSUE #5851: Use new layers selection bottom sheet UI in all map views
It should be possible to add and delete layers from any map view in the app.

### LINKED ISSUE #5852: Use new layers selection bottom sheet UI in settings
It should be possible to add and delete layers from settings (in the Maps section).

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5845 
Closes #5851 
Closes #5852 

#### Why is this the best possible solution? Were any other approaches considered?
Generally, everything has been implemented according to what we had discussed earlier. The only difference might be related to this acceptance criteria:

From #5845
```
Given I'm filling out a form with a select one from map question
And I've copied reference layers to both the global and the current project's directory
When I open the select one from map view
And I press the layers button
And I swipe the bottom sheet up above the middle of the screen
Then the bottom sheet should fill the screen
```
The bottom sheet can be expanded but only if it contains a long list of layers. Otherwise, it's not possible because everything is visible and there is no need to do that. This is the default behavior and I think it doesn't make sense to change it.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
I think it will be better to these the whole feature at once when all the issues are addressed.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@seadowg could you take a look at the test that is failing here please?https://github.com/getodk/collect/blob/b3cd4a057832c48ab04c0c380274e74e4ee28fe2/geo/src/test/java/org/odk/collect/geo/selection/SelectionMapFragmentTest.kt#L407
Here is the reason:
```java
Caused by: java.lang.ClassCastException: class org.odk.collect.geo.support.RobolectricApplication cannot be cast to class org.odk.collect.maps.MapsDependencyComponentProvider
```
The test is in the `geo` module and it opens the list of layers (before it was a dialog now it's a bottom sheet I called `OfflineMapLayersPicker`). `OfflineMapLayersPicker` is in the `maps` module though and I use dagger to inject dependencies there https://github.com/getodk/collect/pull/6118/files#diff-9e6c9d8d539aa94749d2b7fef8bb4c48090de8859f4484a80945ebb71fdff5ffR37 this line causes the problem because I can't cast `org.odk.collect.geo.support.RobolectricApplication` to `org.odk.collect.maps.MapsDependencyComponentProvider`. Do you think solving this is possible? How would you fix this?
You might ask why I didn't use a fragment factory (or a ViewModel) to pass dependencies via constructor so that injecting with dagger wouldn't be needed. The answer is that the new bottom sheet is used in at least three different places so I would need to pass dependencies, I need (there are a few of them) to all those three places and there, use a fragment factory to build my fragment. This doesn't sound easy to maintain so I decided to use dagger. Alternatively, I could have built one fragment factory (with all the dependencies I need) and inject it into those three separate classes which would be easier than injecting all the dependencies I need (3 or 4) separately. But still, that would mean injecting the factory into those three places plus this is not how we have used fragment factories so far.
So basically I still think using dagger to inject dependencies to `OfflineMapLayersPicker` is the best way but can we solve that problem with tests?

**Comment by seadowg** [comment]
> The test is in the geo module and it opens the list of layers (before it was a dialog now it's a bottom sheet I called OfflineMapLayersPicker). OfflineMapLayersPicker is in the maps module though and I use dagger to inject dependencies there https://github.com/getodk/collect/pull/6118/files#diff-9e6c9d8d539aa94749d2b7fef8bb4c48090de8859f4484a80945ebb71fdff5ffR37 this line causes the problem because I can't cast org.odk.collect.geo.support.RobolectricApplication to org.odk.collect.maps.MapsDependencyComponentProvider. Do you think solving this is possible? How would you fix this?

The `RobolectricApplication` in `geo` needs to implement `MapsDependencyComponentProvider` as well as `GeoDependencyComponentProvider` (just like how `Collect` implements both of them). 

I do think that having Dagger dependencies go down multiple layers (`geo` and then `maps`) in the dependency hierarchy feels awkward though (although we've probably already done it somewhere else). One point to make here is that we should really be using a ViewModel in `OfflineMapLayersPicker` instead of talking to `Scheduler` directly as otherwise you'll run into problems with device rotation. 

I think a setup like `MultiSelectListFragment` would be best here: `OfflineMapLayersPicker` takes a ViewModel (or View Model Factory) and `ExternalWebPageHelper` in as constructor args and then the parent Fragment `SelectionMapFragment` passes them via a `FragmentFactory` on its child `FragmentManager`. `SelectionMapFragment` can either construct these (with injected dependencies) or just pass them along from its own constructor. This would be similar to how `MultiSelectListFragment` is used in `DeleteBlankFormFragment`. I think my preference overall is to avoid Dagger in components that don't need them, and that's thankfully now true for Fragments.

Happy to discuss more here on Slack if that's not clear!

**Comment by grzesiek2010** [comment]
@seadowg please take a look if you have a moment when I'm off. https://github.com/getodk/collect/issues/5845 should be ready I wanted to extend this pr a little bit and address https://github.com/getodk/collect/issues/5851 and https://github.com/getodk/collect/issues/5852 as well so I will add some more commits but you can focus on https://github.com/getodk/collect/issues/5845.

File: maps/src/main/res/layout/offline_map_layers_picker.xml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    <View
+        android:id="@+id/bottom_divider"
+        android:layout_width="match_parent"
+        android:layout_height="1dp"
+        android:background="@drawable/list_item_divider"
+        app:layout_constraintBottom_toTopOf="@id/save" />
+
+    <org.odk.collect.androidshared.ui.multiclicksafe.MultiClickSafeMaterialButton
+        android:id="@+id/save"
+        style="?materialButtonStyle"
```
Comment (grzesiek2010) [inline maps/src/main/res/layout/offline_map_layers_picker.xml:123]: @seadowg please take a look at this button. The styling does not work well. If it was an activity everything would be fine (for example we use the same styling in the crash-handler module and there it works well). There is something wrong with using it in fragments in separate modules I guess. Do you maybe know what is wrong?

Comment (seadowg) [reply maps/src/main/res/layout/offline_map_layers_picker.xml:123]: Oh wow that's weird! It looks to me like bottom sheets might get their own theme overlay set on them (although I haven't looked into it deeply). The hint for that is that you can set style to `@style/Widget.Material3.Button` and get what you'd expect, but using `?materialButtonStyle` is just giving us a bolder text button. My guess is that bottom sheets have an overlay theme that overrides `?materialButtonStyle` and that we might need to customise and set our own theme for them (like we do with dialogs). The Material Components repo on Github is probably a good place to look into.

**Review by seadowg — CHANGES_REQUESTED** [review]
https://github.com/getodk/collect/pull/6118#discussion_r1606483903

File: maps/src/main/res/layout/offline_map_layers_picker.xml (hunk 1 above)
Comment (grzesiek2010) [reply maps/src/main/res/layout/offline_map_layers_picker.xml:123]: Yeah, it looks like we need to set our own theme for bottom sheets.

File: geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt [hunk 2]
```diff
@@ -112,4 +108,14 @@ open class GeoDependencyModule {
             }
         }
     }
+
+    @Provides
+    open fun providesOfflineMapLayersPickerViewModelFactory(): OfflineMapLayersPickerViewModel.Factory {
```
Comment (seadowg) [inline geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt]: I think I'd prefer it if we just added providers for `ReferenceLayerRepository` and `SettingsProvider` here. We could then avoid a specific implementation of `ViewModelProvider.Factory` for `OfflineMapLayersPickerViewModel` and just have a shared view model factory for geo components that can construct one (like we do with form entry). This approach keeps a more specific set of components "injectable" (the "building blocks") and should reduce the amount of view model factories we need.

**Review by seadowg — CHANGES_REQUESTED** [review]
I've just done a quick overview of the general structure and had one comment.

Also, I don't have an any mbtiles to try out, but how does the map reload with the new layer? It seems like the underlying setting changes when the user selects a tile, but it's not clear to me what then triggers that being used. I'm probably missing something obvious!

File: geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt (hunk 2 above)
Comment (grzesiek2010) [reply geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt]: > I think I'd prefer it if we just added providers for ReferenceLayerRepository and SettingsProvider here. We could then avoid a specific implementation of ViewModelProvider.Factory for OfflineMapLayersPickerViewModel and just have a shared view model factory for geo components that can construct one (like we do with form entry). 

I see what you mean. I think the reason I did this in that way was that I wanted to have one class to wrap the three dependencies I need (`ReferenceLayerRepository`, `Scheduler`, `SettingsProvider` and potentially more when I add importing/deleting layers) so that I don't have to inject all of them separately in `GeoPolyActivity`, `GeoPointMapActivity`, `SelectionMapFragment` and `MapsPreferencesFragment` (every place where we use the new maps picker). Do you think converting `OfflineMapLayersPickerViewModel.Factory` to a more general one (but still injecting it in the same way) that would in the future as you said work like `FormEntryViewModelFactory` would be a good trade off here?

Comment (seadowg) [reply geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt]: > I think the reason I did this in that way was that I wanted to have one class to wrap the three dependencies I need (ReferenceLayerRepository, Scheduler, SettingsProvider and potentially more when I add importing/deleting layers) so that I don't have to inject all of them separately in GeoPolyActivity, GeoPointMapActivity, SelectionMapFragment and MapsPreferencesFragment (every place where we use the new maps picker).

I think it's ok to have all those injections. I like the model of dependency injection just handling the edges of the system. The hope there is you end up with more being injected into entry point "black boxes" (like Activity objects), but less actual depdencies being managed by the framework with plain ol' constructors doing most of the work.

Maybe a nice way of having our cake and eating it here is to create two `Factory` implementations - one for `geo` (`GeoViewModelFactory`) and one for `MapPreferenecesFragment` in `collect_app` (`MapPreferencesViewModelFactory`). Then we could just use `@Inject` on their constructors. You still only have to deal with one `@Inject` in both places, but we still don't have declare the factory as a dependency of `geo` anywhere. I'd thought about doing this for `FormEntryViewModelFactory`  before, but hadn't got round to it. I like this approach as it gives us the convenience of Dagger in the Activity/Fragment, but doesn't mean managing dependencies outside (in the module etc) and you still have a simple ViewModel factory constructor.

Comment (grzesiek2010) [reply geo/src/main/java/org/odk/collect/geo/DaggerSetup.kt]: I've decided to go for the first option.

**Comment by grzesiek2010** [comment]
> Also, I don't have an any mbtiles to try out, but how does the map reload with the new layer? It seems like the underlying setting changes when the user selects a tile, but it's not clear to me what then triggers that being used. I'm probably missing something obvious!

This is a thing I haven't had to touch even. There is a listener that is called whenever the settings are updated and then we update the map in the map fragment that we use (Google, Mapbox or OSM).

**Comment by seadowg** [comment]
Sorry I really think we should split this up. I've started going through it commit but commit, but because the PR builds things up iteratively (like d2517feea771f732e23fd651b7790d04716ec7a9 for example), I'd really like to be able to use "Files changed" to also keep track of the final state of things so I'm not commenting on intermediarey states. Just to be clear, I really like the step by step approach, it just hard to combine that with a partial review given the tooling we have.

This should be split after d6c9705e12639dbc3a1bc09dc61fde08188a919d - another branch should be created there and this one should be `git reset` to a1945243e69f927c2a39c67bdc357065984ad8c3 and then pushed. I know managing stacked PRs is a pain, but I'm happy to help with the branch management here. I tend to use [`git rebase --onto`](https://www.seadowg.com/2023/01/12/how-to-git-with-github.html#updating-dependent) to deal with stacked branch updates, and I've also been looking at Git Town's [stacked changes] tooling which I think we might want to start adopting as I think we most likely want to be doing this more to cut the size of reviews down.

**Comment by grzesiek2010** [comment]
@seadowg ok the first part is ready.

**Comment by seadowg** [comment]
> @seadowg ok the first part is ready.

Thanks so much! That's going to be really helpful.

File: collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java [hunk 3]
```diff
@@ -569,10 +571,12 @@ public PreferenceVisibilityHandler providesDisabledPreferencesRemover(SettingsPr
     }
 
     @Provides
-    public ReferenceLayerRepository providesReferenceLayerRepository(StoragePathProvider storagePathProvider) {
+    public ReferenceLayerRepository providesReferenceLayerRepository(StoragePathProvider storagePathProvider, SettingsProvider settingsProvider) {
         return new DirectoryReferenceLayerRepository(
-                storagePathProvider.getOdkDirPath(StorageSubdirectory.LAYERS),
-                storagePathProvider.getOdkDirPath(StorageSubdirectory.SHARED_LAYERS)
+                Arrays.asList(storagePathProvider.getOdkDirPath(StorageSubdirectory.LAYERS), storagePathProvider.getOdkDirPath(StorageSubdirectory.SHARED_LAYERS)),
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:576]: I think I'd prefer it if `DirectoryReferenceLayerRepository` wasn't responsible for filter supported lists. It feels like more responsibility than it needs to have.

It look like `#getAllSupported` is only called in `OfflineMapLayersPickerViewModel` so could it not deal with support? It feels more natural to me given it's closer to the UI and the current map engine being used.

File: maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            .forClass(OfflineMapLayersPicker::class) {
+                OfflineMapLayersPicker(referenceLayerRepository, scheduler, settingsProvider, externalWebPageHelper)
+            }.build()
+    )
+
+    @Test
+    fun `clicking the 'cancel' button dismisses the layers picker`() {
+        val scenario = launchFragment()
+
+        scheduler.flush()
```
Comment (seadowg) [inline maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt]: You probably don't want `flush` here as you shouldn't need to wait to hit "Cancel". That goes for a few other tests in here as well.

File: maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+import org.odk.collect.settings.InMemSettingsProvider
+import org.odk.collect.settings.keys.ProjectKeys
+import org.odk.collect.shared.TempFiles
+import org.odk.collect.strings.R.string
+import org.odk.collect.testshared.FakeScheduler
+import org.odk.collect.testshared.RecyclerViewMatcher.Companion.withRecyclerView
+import org.odk.collect.webpage.ExternalWebPageHelper
+
+@RunWith(AndroidJUnit4::class)
+class OfflineMapLayersPickerTest {
```
Comment (seadowg) [inline maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt:39]: You can probably use `EspressoHelpers` throughout here.

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPicker.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import androidx.lifecycle.ViewModelProvider
+import com.google.android.material.bottomsheet.BottomSheetBehavior
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import org.odk.collect.androidshared.ui.GroupClickListener.addOnClickListener
+import org.odk.collect.async.Scheduler
+import org.odk.collect.maps.databinding.OfflineMapLayersPickerBinding
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.webpage.ExternalWebPageHelper
+
+class OfflineMapLayersPicker(
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPicker.kt:19]: Just something to consider: I'd be tempted to pass in the `ViewModel` or `ViewModelProvider.Factory` here instead of the ViewModel's dependencies. That would let you have more concentrated tests of the Fragment and ViewModel. 

I don't think that would be very advantageous in this PR (`OfflineMapsLayerPickerTest` is pretty simple and easy to read), but it might be useful down the line if you wanted to test something like the ordering of layers (which would be easier to test at the ViewModel level).

This is definitely related to our earlier discussion on the injection setup, and I think I'm again suggesting the approach I put forward in https://github.com/getodk/collect/pull/6118#discussion_r1611678238.

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    private val onSelectedLayerChanged: (String?) -> Unit
+) : RecyclerView.Adapter<OfflineMapLayersAdapter.ViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val binding = OfflineMapLayerBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        return ViewHolder(binding)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.binding.radioButton.setChecked(false)
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersAdapter.kt:21]: It looks like the recycling here isn't tested. 

It could be worth extracting a helper adapter for single select lists like this (like we already have for multi select in `MultiSelectAdapter`), but that potentially a follow up change.

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.settings.keys.ProjectKeys
+
+class OfflineMapLayersPickerViewModel(
+    private val referenceLayerRepository: ReferenceLayerRepository,
+    scheduler: Scheduler,
+    private val settingsProvider: SettingsProvider
+) : ViewModel() {
+    private val _data = MutableLiveData<Pair<List<ReferenceLayer>, String?>>()
+    val data: LiveData<Pair<List<ReferenceLayer>, String?>> = _data
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt:16]: It's probably worth creating a specific data class to represent the view data here (rather than using `Pair`) just so it's a little more readable.

File: maps/src/main/res/layout/offline_map_layer.xml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    android:layout_height="wrap_content"
+    android:paddingVertical="@dimen/margin_extra_small">
+
+    <com.google.android.material.radiobutton.MaterialRadioButton
+        android:id="@+id/radio_button"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:minWidth="0dp"
+        android:minHeight="0dp"
+        android:translationX="-5dp"
```
Comment (seadowg) [inline maps/src/main/res/layout/offline_map_layer.xml]: Why is this here?

File: maps/src/main/res/layout/offline_map_layer.xml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:paddingVertical="@dimen/margin_extra_small">
```
Comment (seadowg) [inline maps/src/main/res/layout/offline_map_layer.xml]: Should the padding be internal to the `RadioButton` instead so that the whole view is clickable?

File: maps/src/main/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepository.kt [hunk 11]
```diff
@@ -15,15 +15,15 @@ class DirectoryReferenceLayerRepository(private val directoryPaths: List<String>
...
+            ReferenceLayer(getIdForFile(it.second, it.first), it.first, getName(it.first))
         }.distinctBy { it.id }
     }
 
     override fun get(id: String): ReferenceLayer? {
         val file = getAllFilesWithDirectory().firstOrNull { getIdForFile(it.second, it.first) == id }
 
         return if (file != null) {
-            ReferenceLayer(getIdForFile(file.second, file.first), file.first)
+            ReferenceLayer(getIdForFile(file.second, file.first), file.first, getName(file.first))
```
Comment (seadowg) [inline maps/src/main/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepository.kt:24]: I don't think there's a test for the name being added here.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: maps/src/main/res/layout/offline_map_layer.xml (hunk 9 above)
Comment (grzesiek2010) [reply maps/src/main/res/layout/offline_map_layer.xml]: The `RadioButton` has a margin that I remove using this. I don't want it. I want the button to start from the edge of the layout without any extra margin. The `TriggerWIdget` uses the same.

File: collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:576]: > It look like #getAllSupported is only called in OfflineMapLayersPickerViewModel so could it not deal with support?

That's right `#getAllSupported` is called only in that one place but on the other hand `#getAll` is only called internally by `#getAllSupported`. There is no real use case for returning all layers as we would need to filter them either way.

I think I would rather get rid of `#getAll` at all and maybe rename `#getAllSupported` to `#getAll` but keep the filtering (the name does not have to reflect that fact I think, tests is enough).

Does that make sense?

File: maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt]: Done.

File: maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt (hunk 5 above)
Comment (grzesiek2010) [reply maps/src/test/java/org/odk/collect/maps/layers/OfflineMapLayersPickerTest.kt:39]: Done. I've also added more tests that I already had in the previous version of this pr but the commits they were added by have been removed.

File: maps/src/main/res/layout/offline_map_layer.xml (hunk 10 above)
Comment (grzesiek2010) [reply maps/src/main/res/layout/offline_map_layer.xml]: Right, it would be better if the padding was internal. Done.

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt (hunk 8 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt:16]: I agree and I came to the same conclusion but at a later stage (while working on deleting layers when I need even more details like for example if the layer item is expanded or collapsed). If it's ok, let's maybe keep it unchanged for now so that I can cherry-pick commits that will improve this code without any conflicts (in one of the next prs).

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPicker.kt (hunk 6 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPicker.kt:19]: Let's keep this in mind but do not change now. I'm not against but I also think it's not something we definitely need to change.
When it comes to testing I already verify the order of displayed layers in `OfflineMapLayersPickerTest`.
Even if after merging this pr and creating a new one I change my mind and start thinking that yeah we should have it, it would be better to do that later so that I can cherry-pick the commits I've removed easier.

File: maps/src/main/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepository.kt (hunk 11 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepository.kt:24]: Done in https://github.com/getodk/collect/pull/6118/commits/901a9ab2646dfe4720ad7ce33c7ca7d82fa9ffd0

File: collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:576]: Yeah I think I like this approach especially as I can also use the same dependency (`mapConfigurator`) to read the file name see: https://github.com/getodk/collect/pull/6118/commits/901a9ab2646dfe4720ad7ce33c7ca7d82fa9ffd0
It looks like this: https://github.com/getodk/collect/pull/6118/commits/a6c52cb211fb1304f0f1d646f93a1c984c67c0f6

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersAdapter.kt (hunk 7 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersAdapter.kt:21]: Right, I wanted to avoid creating a custom view for list elements but that would make testing easier. Let's wait until it has its final form (expandable with the delete button).

File: collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java (hunk 3 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:576]: Yeah that makes sense!

File: maps/src/main/res/layout/offline_map_layer.xml (hunk 9 above)
Comment (seadowg) [reply maps/src/main/res/layout/offline_map_layer.xml]: Ah I see now: `offline_map_layers_picker.xml` defines a guideline that the `RecyclerView` starts at so we don't need the built in margin in here. The problem there is that we end up with the start edge of the screen not being tappable, even though I'd expect the list item to be selectable across the full horizontal space. I think we should make the `RecyclerView` full width and remove this (although we might want to adjus the start padding to one of our standard margins).

**Review by seadowg — CHANGES_REQUESTED** [review]
https://github.com/getodk/collect/pull/6118#discussion_r1624075784

File: maps/src/main/res/layout/offline_map_layer.xml (hunk 9 above)
Comment (grzesiek2010) [reply maps/src/main/res/layout/offline_map_layer.xml]: I mean not so long ago we decided not to make the `RecyclerView` responsible for displaying choices in `select_one`/`select_multiple` questions full-width and I think that one is much more important so I thought it's ok. Do you really want it to be the way you described it? If so I'm ok to do that but to me it's not an issue.

Comment (seadowg) [reply maps/src/main/res/layout/offline_map_layer.xml]: I think that was more to do with how we controlled widget margins. I don't think we need to do anything special here and can just let the `RecyclerView` be full width.

**Review by seadowg — CHANGES_REQUESTED** [review]
https://github.com/getodk/collect/pull/6118#discussion_r1624399545

File: maps/src/main/res/layout/offline_map_layer.xml (hunk 9 above)
Comment (grzesiek2010) [reply maps/src/main/res/layout/offline_map_layer.xml]: Ok, done.

File: androidshared/src/main/res/drawable/radio_button_inset.xml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<inset xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (seadowg) [inline androidshared/src/main/res/drawable/radio_button_inset.xml:2]: It's super confusing that you have to do this and can't just set some kind of padding on the `MaterialRadioButton`. Could you file an issue with this example at https://github.com/material-components/material-components-android? That really needs to get fixed.

File: maps/src/test/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepositoryTest.kt [hunk 13]
```diff
@@ -4,40 +4,71 @@ import org.hamcrest.MatcherAssert.assertThat
...
+    fun getAll_returnsAllSupportedLayersInTheDirectory() {
         val dir = TempFiles.createTempDir()
         val file1 = TempFiles.createTempFile(dir)
         val file2 = TempFiles.createTempFile(dir)
-
-        val repository = DirectoryReferenceLayerRepository(dir.absolutePath)
-        assertThat(repository.getAll().map { it.file }, containsInAnyOrder(file1, file2))
+        val file3 = TempFiles.createTempFile(dir)
+        val mapConfigurator = mock<MapConfigurator>().also {
+            whenever(it.supportsLayer(file1)).thenReturn(true)
```
Comment (seadowg) [inline maps/src/test/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepositoryTest.kt]: It might be a little more readable (for each test) to write a stub outside of Mockito here. For example:

```kotlin
private class StubMapConfigurator : MapConfigurator {

    private val files = mutableMapOf<File, Pair<Boolean, String>>()

    override fun supportsLayer(file: File?): Boolean {
        return files[file]!!.first
    }

    override fun getDisplayName(file: File?): String {
        return files[file]!!.second
    }

    fun addFile(fike: File, displayName: String, isSupported: Boolean) {

    }

    override fun isAvailable(context: Context?): Boolean {
        TODO("Not yet implemented")
    }

    override fun showUnavailableMessage(context: Context?) {
        TODO("Not yet implemented")
    }

    override fun createPrefs(context: Context?, settings: Settings?): MutableList<Preference> {
        TODO("Not yet implemented")
    }

    override fun getPrefKeys(): MutableCollection<String> {
        TODO("Not yet implemented")
    }

    override fun buildConfig(prefs: Settings?): Bundle {
        TODO("Not yet implemented")
    }
}
```

**Review by seadowg — CHANGES_REQUESTED** [review]

File: maps/src/test/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepositoryTest.kt (hunk 13 above)
Comment (grzesiek2010) [reply maps/src/test/java/org/odk/collect/maps/layers/DirectoryReferenceLayerRepositoryTest.kt]: Done.

File: androidshared/src/main/res/drawable/radio_button_inset.xml (hunk 12 above)
Comment (grzesiek2010) [reply androidshared/src/main/res/drawable/radio_button_inset.xml:2]: Ok https://github.com/material-components/material-components-android/issues/4198

**Review by seadowg — APPROVED** [review]

**Comment by seadowg** [comment]
@getodk/testers going to skip "needs testing" here so you can test the whole feature once managing tiles is added.

File: maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt (hunk 8 above)
Comment (grzesiek2010) [reply maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersPickerViewModel.kt:16]: Done.
